### PR TITLE
`:doc`: support `__functor`

### PR DESCRIPTION
**Pull request by roberth**
# Motivation

Render more docs!

# Context

- Builds upon #11072 

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: tests/functional/repl/doc-functor.nix [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    override = throw "not implemented";
+  };
+
+  helper2 = makeVeryOverridable square;
+
+  # The RFC might be ambiguous here. The doc comment from makeVeryOverridable
+  # is "inner" in terms of values, but not inner in terms of expressions.
+  # Returning the following attribute comment might be allowed.
+  # TODO: I suppose we could look whether the attribute value expression
+  #       contains a doc, and if not, return the attribute comment anyway?
```
Comment (roberth) [inline tests/functional/repl/doc-functor.nix:55]: @hsjobeki What do you think?
Currently we follow the [disambiguation rule](https://github.com/NixOS/rfcs/pull/145/files#diff-5050f8573da95e8531bf14e0b0de89d58a680478678ed4f2ed20f9e53def4dd9R200) at the value level.
Was that the intent, or should we only apply it at the expression level, so that in cases like this we can resolve to a more specific comment.
Or perhaps more simply, when the `:doc` query is specifically for an attribute, perhaps we should just prefer the attribute-based lookup?
I'm also considering the option that we should just show everything and treat the rule as a preference or priority.

> ```
> nix-repl> :doc lib.helper3
> Function `__functor`\
>   … defined at /path/to/tests/functional/repl/doc-functor.nix:45:23
> 
> 
> This is a function that can be overridden.
> ```

Comment (hsjobeki) [reply tests/functional/repl/doc-functor.nix:55]: > Or perhaps more simply, when the :doc query is specifically for an attribute, perhaps we should just prefer the attribute-based lookup?

Yes. But functors are polymorph. I would show (maybe concatenate) both, or add an option for the user to decide what he is interested in.

As a user i would expect `:doc lib.helper3` to yield.

```
:doc lib.helper3
->  Compute x^3
```

The property that the function is wrapped with `makeVeryOverridable` is not very interesting.
But i dont know how to solve this issue yet.

Comment (hsjobeki) [reply tests/functional/repl/doc-functor.nix:55]: For example this one edge case that i couldn't solve on a value level, because lamdbas are opaque.

`pkgs.fetchFromGitHub`
->
https://github.com/nixos/nixpkgs/blob/master/lib/customisation.nix#L134:C17

**Comment by hsjobeki** [comment]
Great PR! This is exactly what noogle does with functors. I can confirm it works for (most) functors in nixpkgs.
But has some limitations as discussed.

**Comment by hsjobeki** [comment]
Something i didnt really dig into yet IS the c-api. I could Imagine that someone could build a documentation Generator for nix Code as Independent Project

**Comment by roberth** [comment]
None of this is exposed in the C API yet. We might need a tier system or something for its interface stability when we add very new things like this, because I don't feel confident that we know what the interface should look like, particularly because the availability of features and their behavior is in part driven by evaluator implementation details. This may or may not be a problem, but it feels premature to make promises about it.

**Review by tomberek — APPROVED** [review]

**Comment by hsjobeki** [comment]
@roberth I have a partial solution for the `lib.makeOverridable` problem. 

If the wrapped function uses a pattern  (`{ rev, ... }@args`)

Then we can introspect them to find the pattern position. The lambda is just the next lambda node up in the ast.


```nix
pkgs.fetchFromGitHub.__functionArgs.sparseCheckout
```

The tracked attribute position:

```json
 "/nix/store/sm63abmh01jvz9i31xsakc58ys1dmvcp-source/pkgs/build-support/fetchgithub/default.nix#L7:C3",
 ```
